### PR TITLE
Concurrent remembered set scanning

### PR DESCRIPTION
**Pull request by kdnilsen**
This pull request includes code to perform remembered set scanning concurrently with ongoing mutator execution.  As currently implemented, the remembered set is copied to a read-only remembered set during init-mark safepoint.  After copying, all cards in the remembered set are restored to a clean value.  Following the init-mark safepoint, mutator thread write barriers update the active remembered set copy every time a reference field is overwritten by dirtying the corresponding cards.

Concurrent remembered set scanning for the purpose of marking works with the read-only copy of the remembered set.  When concurrent remembered set scanning encounters an interesting pointer, the corresponding card of the active remembered set is overwritten dirty.

The remembered set is scanned a second time concurrently during the update-refs phase of gc.  The second scan uses the same read/write active copy of the remembered set that is being overwritten by the mutator write barrier.

The SATB barrier and load reference barriers are sufficient to resolve any races between setting of dirty flags and scanning of the remembered set.

A possible future enhancement involves replacing the remembered set copy (during init-mark safepoint) with a simple pointer swap operation to exchange the roles of a read-byte-map and a write-byte-map.  Making this change is deferred because it requires far-reaching changes to the implementations of mutator write barriers.  As currently implemented, these barriers hard-code the address of the card-table byte map.

This patch passes tier1 and hotspot_gc_shenandoah regression tests.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/shenandoah pull/37/head:pull/37` \
`$ git checkout pull/37`

Update a local copy of the PR: \
`$ git checkout pull/37` \
`$ git pull https://git.openjdk.java.net/shenandoah pull/37/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 37`

View PR using the GUI difftool: \
`$ git pr show -t 37`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/shenandoah/pull/37.diff">https://git.openjdk.java.net/shenandoah/pull/37.diff</a>

</details>

### CODE REVIEW

**Comment by kdnilsen** [comment]
The merge from head appears to have introduced a bug into the implementation.  I am investigating.

**Comment by kdnilsen** [comment]
Even after the merge, I still pass tier1 and hotspot_gc_shenandoah regression tests.  However, my extremem stress test is experiencing a sporadic SIGSEGV crash in Shenandoah::mark_loop_work()

**Comment by zhengyu123** [comment]
Would you mind to merge with master? it helps the review.

Thanks.

**Comment by kdnilsen** [comment]
I think the code is now merged with master.  But my initial commits were not.  Does it look merged to you?

**Comment by rkennke** [comment]
It looks better now. If you think it's ready for review, then please push the button 'Ready for review' to lift the Draft status.

**Comment by kdnilsen** [comment]
There are a couple of loose ends I'm trying to tie up before I change to RFR.  Thanks much.

**Comment by rkennke** [comment]
Please check and fix jcheck problems: https://github.com/openjdk/shenandoah/pull/37/checks?check_run_id=2658483300

**Comment by kdnilsen** [comment]
Testing with the latest commit on this branch seems to reveal a bug in Full GC relocation of humongous objects.  The error is intermittently reproduced with the following command line:
 make exploded-test TEST=test/hotspot/jtreg/gc/shenandoah/oom/TestAllocLargeObj.java CONF=fast \
  JTREG="JOBS=4;TIMEOUT_FACTOR=8;JAVA_OPTIONS=-Djavatest.maxOutputSize=300000 -XX:+UseShenandoahGC \
         -XX:ShenandoahGCMode=generational -XX:-RegisterReferences -Xlog:gc" >LargeAllocObj.out 2>LargeAllocObj.err
         
To see "proof" of the error condition, modify the source to "#define KELVIN_DEBUG_LIVENESS" at each spot that currently #undef'ines this symbol.  Then run the same test.

Would appreciate a second set of eyes on this if anyone has time to take a look.  I've not yet studied the implementation of Full GC as Generational Shenandoah does not yet support it.  (This error appears to manifest because some other jtreg parameter seems to be overriding the request to run under generational ShenandoahGCMode.)

**Comment by rkennke** [comment]
> Testing with the latest commit on this branch seems to reveal a bug in Full GC relocation of humongous objects. The error is intermittently reproduced with the following command line:
> make exploded-test TEST=test/hotspot/jtreg/gc/shenandoah/oom/TestAllocLargeObj.java CONF=fast
> JTREG="JOBS=4;TIMEOUT_FACTOR=8;JAVA_OPTIONS=-Djavatest.maxOutputSize=300000 -XX:+UseShenandoahGC
> -XX:ShenandoahGCMode=generational -XX:-RegisterReferences -Xlog:gc" >LargeAllocObj.out 2>LargeAllocObj.err
> 
> To see "proof" of the error condition, modify the source to "#define KELVIN_DEBUG_LIVENESS" at each spot that currently #undef'ines this symbol. Then run the same test.
> 
> Would appreciate a second set of eyes on this if anyone has time to take a look. I've not yet studied the implementation of Full GC as Generational Shenandoah does not yet support it. (This error appears to manifest because some other jtreg parameter seems to be overriding the request to run under generational ShenandoahGCMode.)

I'm taking a look.

**Comment by rkennke** [comment]
The test seems to completely override the command line options. This reproduces fairly reliably here:
./build/linux-x86_64-server-fastdebug/images/jdk/bin/java -cp $PATH_TO_TestAllocLargeObj_class -Xmx16m -XX:+UnlockExperimentalVMOptions -XX:+UseShenandoahGC TestAllocLargeObj test
From the looks of it, it encounters a humongous continuation in a FREE region that should not be in FREE, and it happens after a fullgc. I'll look deeper later.

**Comment by rkennke** [comment]
Kelvin, might this be a race of some sorts? I am getting a failure that region 57 is FREE even though it should not. Log shows:

[0,142s][debug][gc                   ] NOT clearing bitmap for Humongous region [0x00000000ffdc0000, 0x00000000ffe00000], top_bitmap: 0x00000000ffe00000 at transition from FREE to YOUNG
[0,142s][trace][gc                   ] Changing affiliation of region 56 from FREE to YOUNG
[0,142s][debug][gc                   ] Capturing TAMS for YOUNG Region 56, was: ffe00000, now: ffe00000

[0,142s][debug][gc                   ] NOT clearing bitmap for Humongous region [0x00000000ffe00000, 0x00000000ffe40000], top_bitmap: 0x00000000ffe40000 at transition from FREE to YOUNG
[0,142s][trace][gc,tlab              ] ThreadLocalAllocBuffer::compute_size(524290) returns failure
[0,142s][info ][gc                   ] Failed to allocate Shared, 4096K
[0,142s][info ][gc                   ] Cancelling GC: Allocation Failure

It looks like it's getting an alloc failure *while* setting up region 57, and does not get to the point where it sets the affiliation of region 57.
Reproduces reliably with:

./build/linux-x86_64-server-fastdebug/images/jdk/bin/java -cp build/linux-x86_64-server-fastdebug/test-support/jtreg_test_hotspot_jtreg_gc_shenandoah_oom_TestAllocLargeObj_java/classes/0/gc/shenandoah/oom/TestAllocLargeObj.d -Xmx16m -XX:+UnlockExperimentalVMOptions -XX:+UseShenandoahGC -Xlog:gc*=trace:file=gc.log TestAllocLargeObj test

**Comment by kdnilsen** [comment]
There may be a race of some sort.  In theory, when we change affiliation from FREE to YOUNG, we will initialize the relevant fields when we make that set_affiliation() call.  But region 57 does not get changed by full GC.  Only region 56 gets changed.

**Comment by kdnilsen** [comment]
I've determined that the "bug" is in ShenandoahFullGC::compact_humongous_objects(), near the invocation of Copy::aligned_conjoint_words().

This code needs to update region affiliations, setting each destination region to have the same affiliation as the region from which data is being copied. And it needs to set affiliation to FREE for any region that is completely evacuated by the compacting copy.

It's a bit tricky because the destination regions may overlap with the origin regions.  Too late for me to try to fix this tonight, but I'll resume work tomorrow US time.  Unless someone else wants to tackle this during the European day.

**Comment by kdnilsen** [comment]
I should add that there's a bit more at play here than just setting the region affiliations.  We also need to do the right thing with regards to clearing bitmaps, update-watermarks, and TAMS values for any region that becomes non-free.

**Comment by kdnilsen** [comment]
We do still see occasional crashes and assertion failures, but they are much more rare now.  This branch may be as stable as our generational shenandoah master branch.  Would appreciate a review and feedback.   Thanks.

**Comment by kdnilsen** [comment]
We continue to sort through the rare crashes we still observe.  We have confirmed that several of these crashes are related to weak reference processing and weak root processing.  Those issues are known and fixes are being developed on a different branch.

**Review by rkennke — COMMENTED** [review]
Great work! I only have some comments on cosmetic issues.

File: src/hotspot/share/gc/shenandoah/shenandoahCardTable.cpp [hunk 1]
```diff
@@ -48,3 +81,37 @@ bool ShenandoahCardTable::is_dirty(MemRegion mr) {
 void ShenandoahCardTable::clear() {
   CardTable::clear(_whole_heap);
 }
+
+static bool already_initialized = false;
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahCardTable.cpp:85]: Removed in next commit.

File: src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.cpp [hunk 2]
```diff
@@ -89,7 +89,6 @@ void ShenandoahCollectionSet::add_region(ShenandoahHeapRegion* r) {
   _garbage += r->garbage();
   _used += r->used();
   _has_old_regions |= r->is_old();
-
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.cpp:92]: Fixed in next commit.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 3]
```diff
@@ -514,8 +508,7 @@ void ShenandoahControlThread::resume_concurrent_old_cycle(ShenandoahGeneration*
   // precisely where the regulator is allowed to cancel a GC.
   ShenandoahOldGC gc(generation, _allow_old_preemption);
   if (gc.collect(cause)) {
-    // Old collection is complete, the young generation no longer needs this
-    // reference to the old concurrent mark so clean it up.
+    // Old collection is complete.  The young generation no longer needs this reference to the old concurrent mark so clean it up.
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:511]: Replaced in next commit.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 4]
```diff
@@ -413,11 +413,7 @@ void ShenandoahControlThread::service_concurrent_normal_cycle(
 
   switch (generation) {
     case YOUNG: {
-      // Run a young cycle. This might or might not, have interrupted an ongoing
-      // concurrent mark in the old generation. We need to think about promotions
-      // in this case. Promoted objects should be above the TAMS in the old regions
-      // they end up in, but we have to be sure we don't promote into any regions
-      // that are in the cset (more of an issue for Milestone-8 to worry about).
+      // Run a young cycle. This might or might not, have interrupted an ongoing concurrent mark in the old generation.
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:416]: Restored the comment to its original format, removing mention of Milestone-8 as we are well beyond that now.

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp [hunk 5]
```diff
@@ -520,7 +576,7 @@ void ShenandoahHeapRegion::oop_iterate_humongous(OopIterateClosure* blk) {
 }
 
 ShenandoahHeapRegion* ShenandoahHeapRegion::humongous_start_region() const {
-  ShenandoahHeap* heap = ShenandoahHeap::heap();
+ ShenandoahHeap* heap = ShenandoahHeap::heap();
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:579]: Restored original formatting.

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp]: Going to do one more commit on this, to remove extraneous space character.

**Review by rkennke — APPROVED** [review]
Looks good now! Thank you!

**Pull request closed without merging**